core/field_variable.js

```diff
@@ -96,39 +122,79 @@ Blockly.FieldVariable.prototype.setSourceBlock = function(block) {
 };
 
 /**
- * Get the variable's name (use a variableDB to convert into a real name).
- * Unline a regular dropdown, variables are literal and have no neutral value.
- * @return {string} Current text.
+ * Get the variable's ID.
+ * @return {string} Current variable's ID.
  */
 Blockly.FieldVariable.prototype.getValue = function() {
-  return this.getText();
+  return this.variable_ ? this.variable_.getId() : null;
 };
 
 /**
- * Set the variable name.
- * @param {string} value New text.
+ * Get the text from this field.
+ * @return {string} Current text.
  */
-Blockly.FieldVariable.prototype.setValue = function(value) {
-  var newValue = value;
-  var newText = value;
+Blockly.FieldVariable.prototype.getText = function() {
+  return this.variable_ ? this.variable_.name : '';
+};
 
-  if (this.sourceBlock_) {
-    var variable = this.sourceBlock_.workspace.getVariableById(value);
-    if (variable) {
-      newText = variable.name;
-    }
-    // TODO(marisaleung): Remove name lookup after converting all Field Variable
-    //     instances to use ID instead of name.
-    else if (variable = this.sourceBlock_.workspace.getVariable(value)) {
-      newValue = variable.getId();
-    }
-    if (Blockly.Events.isEnabled()) {
-      Blockly.Events.fire(new Blockly.Events.BlockChange(
-          this.sourceBlock_, 'field', this.name, this.value_, newValue));
+/**
+ * Get the variable model for the selected variable.
+ * Not guaranteed to be in the variable map on the workspace (e.g. if accessed
+ * after the variable has been deleted).
+ * @return {?Blockly.VariableModel} the selected variable, or null if none was
+ *     selected.
+ * @package
+ */
+Blockly.FieldVariable.prototype.getVariable = function() {
+  return this.variable_;
+};
+
+/**
+ * Set the variable ID.
+ * @param {string} id New variable ID, which must reference an existing
+ *     variable.
+ */
+Blockly.FieldVariable.prototype.setValue = function(id) {
+  var workspace = this.sourceBlock_.workspace;
+  var variable = Blockly.Variables.getVariable(workspace, id);
+
+  if (!variable) {
+    throw new Error('Variable id doesn\'t point to a real variable!  ID was ' +
+        id);
+  }
+  // Type checks!
+  var type = variable.type;
+  if (!this.typeIsAllowed_(type)) {
+    throw new Error('Variable type doesn\'t match this field!  Type was ' +
+        type);
+  }
+  if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
+    var oldValue = this.variable_ ? this.variable_.getId() : null;
+    Blockly.Events.fire(new Blockly.Events.BlockChange(
+        this.sourceBlock_, 'field', this.name, oldValue, variable.getId()));
```

Comment:
nit: Could use id instead of variable.getId()

Reply:
For reasons I don't remember we made the variable ID private and always access it with getIde().  It doesn't actually change anything, but for convention I'm not accessing private variables directly.

Reply:
but you use "id" directly three lines later. The local method scoped one, not the class field.

Reply:
Done.